### PR TITLE
Suppress warnings in LLVM wrapper when targeting MSVC

### PR DESCRIPTION
**Pull request by sivadeilra**
The LLVM header files generate many warnings when compiled using MSVC. This makes it difficult to work on the LLVM wrapper code, because the warnings and errors that are relevant to local edits are obscured by the hundreds of lines of warnings from the LLVM Headers.

### CODE REVIEW

File: compiler/rustc_llvm/llvm-wrapper/Linker.cpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#include "SuppressLLVMWarnings.h"
```
Comment (lqd) [inline compiler/rustc_llvm/llvm-wrapper/Linker.cpp:1]: drive-by comment: shouldn't this be unnecessary since this file also includes `LLVMWrapper.h`?

Comment (sivadeilra) [reply compiler/rustc_llvm/llvm-wrapper/Linker.cpp:1]: It does, but only after already including several other header files.  The `llvm-c` headers are fine, but `llvm/Linker/Linker.h` is a C++ header, and so could have the same issues.

**Comment by cuviper** [comment]
I don't have much context with MSVC to know if these suppressions are appropriate -- maybe @ChrisDenton can comment?

**Comment by sivadeilra** [comment]
I'm using VS 2022.  The warnings are displayed only when compilation fails, because the warnings are treated as stdout of a `build.rs` build script.  So in normal builds (where everything succeeds), this is not causing a problem.

But I'm doing development on the C++ code in rust_llvm right now.  So whenever I build something and it fails due to a 1-line mistake on my part, I have to scroll through hundreds of lines of irrelevant warnings just to find the one line that is relevant to my work.  I know, I could use a regex or something, but that just adds one more step to a workflow.

**Comment by ChrisDenton** [comment]
Sorry for my slow response. These suppressions do look reasonable to me. It'd be great to actually fix the warnings properly (or else make suppression more targeted) but that seems like an upstream issue? Or if not then it'd be a lot of work.

**Comment by cuviper** [comment]
> The warnings are displayed only when compilation fails, because the warnings are treated as stdout of a `build.rs` build script. So in normal builds (where everything succeeds), this is not causing a problem.

Ah, okay, that build-script hiding is the part I forgot.

> It'd be great to actually fix the warnings properly (or else make suppression more targeted) but that seems like an upstream issue?

Can we target it with pragma push/pop around the includes?

**Comment by sivadeilra** [comment]
> Can we target it with pragma push/pop around the includes?

I tried that, and I found that template instantiations triggered some of these warnings, even outside of the push scope.  :(  It would also require putting the push/pop in a lot more places.

I think the real fixes should happen upstream in LLVM, but I'm not sure whether they'll take those changes since they're super focused on self-hosting Clang.  They make sure that Clang compiles and works with MSVC, but they're fairly slow to take fixes that don't affect Clang.

**Comment by cuviper** [comment]
> > Can we target it with pragma push/pop around the includes?
> 
> I tried that, and I found that template instantiations triggered some of these warnings, even outside of the push scope. :( It would also require putting the push/pop in a lot more places.

Thanks for trying! Then I think it's ok to accept this for now, and maybe future upstream fixes will let us tighten up again.

@bors r+

> I think the real fixes should happen upstream in LLVM, but I'm not sure whether they'll take those changes since they're super focused on self-hosting Clang. They make sure that Clang compiles and works with MSVC, but they're fairly slow to take fixes that don't affect Clang.

If they are true problems, then I hope they'll be accepted -- but it may well be hard work to argue that the MSVC toolchain is right about these cases, and perhaps that Clang is too lax. Or it could just be conditional code that's only broken in the MSVC path, which should take less justification to clean up. (I'm guessing blindly here. :see_no_evil:)

**Comment by sivadeilra** [comment]
Thanks, much appreciated!
